src/typechecker/FStar.TypeChecker.Rel.fs

```diff
@@ -3860,22 +3933,137 @@ let resolve_implicits' env must_total forcelax g =
                until_fixpoint (g'.implicits@out, true) tl in
   {g with implicits=until_fixpoint ([], false) g.implicits}
 
-let resolve_implicits env g = resolve_implicits' env (not env.phase1 && not env.lax)  false g
+let resolve_implicits env g =
+    if Env.debug env <| Options.Other "ResolveImplicitsHook"
+    then BU.print1 "//////////////////////////ResolveImplicitsHook: resolve_implicits////////////\n\
+                    guard = %s\n"
+                    (guard_to_string env g);
+  resolve_implicits' env (not env.phase1 && not env.lax) false g
+
 let resolve_implicits_tac env g = resolve_implicits' env false true  g
 
 let force_trivial_guard env g =
-    let g = solve_deferred_constraints env g |> resolve_implicits env in
+    if Env.debug env <| Options.Other "ResolveImplicitsHook"
+    then BU.print1 "//////////////////////////ResolveImplicitsHook: force_trivial_guard////////////\n\
+                    guard = %s\n"
+                    (guard_to_string env g);
+
+    let solve_goals_with_tac (imps:implicits) (tac:sigelt) =
+        let deferred_goals = DeferredImplicits.sort_goals env imps in
+        let guard = { g with implicits = imps; deferred_to_tac=[] } in
+        let resolve_tac =
+          match tac.sigel with
+          | Sig_let (_, [lid]) ->
+            let qn = Env.lookup_qname env lid in
+            let fv = S.lid_as_fv lid (Delta_constant_at_level 0) None in
+            let dd =
+              match Env.delta_depth_of_qninfo fv qn with
+              | Some dd -> dd
+              | None -> failwith "Expected a dd"
+            in
+            let term = S.fv_to_tm (S.lid_as_fv lid dd None) in
+            term
+          | _ -> failwith "Resolve_tac not found"
+        in
+        let env = { env with enable_defer_to_tac = false } in
+        env.try_solve_implicits_hook env resolve_tac deferred_goals
+    in
+    let solve_deferred_to_tactic g =
+      let deferred = g.deferred_to_tac in
+      match deferred with
+      | [] -> g
+      | _ ->
+        let prob_as_implicit (reason, prob)
+          : implicit * sigelt =
+          match prob with
+          | TProb tp when tp.relation=EQ ->
+            let env = {env with gamma=tp.logical_guard_uvar.ctx_uvar_gamma} in
+            let env_lax = {env with lax=true; use_bv_sorts=true} in
+            let _, tlhs, _ = env.type_of env_lax tp.lhs in
+            let goal_ty = U.mk_eq2 (env.universe_of env_lax tlhs) tlhs tp.lhs tp.rhs in
+            let goal, ctx_uvar, _ =
+              Env.new_implicit_var_aux reason tp.lhs.pos env goal_ty Allow_untyped None
```

Comment:
Should this be `Allow_untyped` here? (Trying to understand why not `Strict`, could `Allow_untyped` be unsound.)

Reply:
I don't think it matters much, since this will be dispatched to a tactic anyway. And in the case of tactics, we do not re-typecheck solutions to goals. So, I left it as Allow_untyped for that reason, since in other usages, Strict means that we re-typecheck solutions.

Reply:
I made it Strict